Add more tests to check mrow fallback when computing preferred logical widths

icesfont opened the pull request:
We currently check that certain invalid MathML markup falls back to being laid out as an `<mrow>` element. Add further tests to check that the preferred logical widths also fallback to being computed in the same way as an `<mrow>` element.

icesfont committed:
Refactor code for generating invalid MathML markup

Since the logic for generating invalid MathML markup will be also used by future tests for testing logical preferred widths, computed baseline, etc. move this into a separate file.

fred-wang commented on mathml/presentation-markup/mrow/mrow-fallback-002.html:
```diff
@@ -0,0 +1,151 @@
+  <div id="log"></div>
+  <div id="container"></div>
+  <div id="invalidMultiscripts">
+    <div style="background: lightgreen; display: inline-block" class="invalid-mmultiscripts-container" data-description="first in-flow child is an <mprescripts>">
```
I believe background: lightgreen is not necessary but maybe that's useful for developers debugging the tests. Up to you but please make it consistent with `wrapShrink()`. Also better to use a more neutral color such as gray as green often means PASS in WPT.

icesfont replied:
Removed in `invalid-markup.js`

fred-wang commented on mathml/presentation-markup/mrow/mrow-fallback-002.html:
```diff
@@ -0,0 +1,151 @@
+               getElementsByClassName("invalid-mmultiscripts-container")).forEach(element => {
+      let reference = element.nextElementSibling;
+      let description = element.dataset.description;
+      console.log(reference, description)
```
please remove console.log

icesfont replied:
Removed

fred-wang commented on mathml/presentation-markup/mrow/mrow-fallback-002.html:
```diff
@@ -0,0 +1,151 @@
+  setup({ explicit_done: true });
+  window.addEventListener("load", () => { loadAllFonts().then(runTests); });
+
+  function isValid(tagName, mspaceCount) {
```
I think this is duplicating a lot of code from `mrow-fallback-002.html`. Can you please factor out some code into a `mathml/presentation-markup/mrow/support/invalid-markup.js` file? I guess that would be `isValid()` and the code generating the invalid markup.

icesfont replied:
Done

icesfont committed:
Add more tests to check mrow fallback when computing preferred logical widths

We currently check that certain invalid MathML markup falls back to being laid out as an `<mrow>` element. Add further tests to check that the preferred logical widths also fallback to being computed in the same way as an `<mrow>` element.

fred-wang commented on mathml/presentation-markup/mrow/mrow-fallback-001.html:
```diff
@@ -0,0 +1,46 @@
+      let reference = element.nextElementSibling;
+      let description = element.dataset.description;
+      test(function() {
+        compareLayout(element, reference, epsilon);
```
can you please also add a test for the baseline alignment here?

Probably

`assert_approx_equals(elementWrapper.getBoundingClientRect().top, referenceWrapper.getBoundingClientRect().top, epsilon);`

would be enough.

icesfont replied:
Added (without the wrapper)

fred-wang reviewed (changes requested):
Thanks. This LGTM, just with the baseline check added.

icesfont committed:
Add tests for computing first line baseline

As part of testing that the layout falls back to that of an `mrow`, also test that the first line baseline falls back to that of an `mrow`.
